### PR TITLE
Handle null values during yaml metadata parsing validation

### LINKED ISSUE #104021: _CastError: Null check operator used on a null value at _validateMetadataMap
On stable/3.0
```
Thread 0 MAIN THREAD _CastError: Null check operator used on a null value
at _validateMetadataMap	(flutter_project_metadata.dart:58)
at new FlutterProjectMetadata	(flutter_project_metadata.dart:99)
at CreateBase.determineTemplateType	(create_base.dart:221)
at CreateCommand._getProjectType	(create.dart:170)
at CreateCommand.runCommand	(create.dart:210)
at FlutterCommand.verifyThenRunCommand	(flutter_command.dart:1342)
```

#99092

### PR DESCRIPTION
**Pull request by jmagman**
And other casting fixups.

Fixes https://github.com/flutter/flutter/issues/104021

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter_tools/test/general.shard/flutter_project_metadata_test.dart [hunk 1]
```diff
@@ -50,6 +50,21 @@ void main() {
     expect(logger.traceText, contains('.metadata file at .metadata was empty or malformed.'));
   });
 
+  testWithoutContext('projectType is populated when version is null', () {
```
Comment (jmagman) [inline packages/flutter_tools/test/general.shard/flutter_project_metadata_test.dart:53]: This test crashes with `_CastError: Null check operator used on a null value` on master.

File: packages/flutter_tools/lib/src/flutter_project_metadata.dart [hunk 2]
```diff
@@ -51,21 +51,18 @@ FlutterProjectType? stringToProjectType(String value) {
...
+  bool _validateMetadataMap(YamlMap map, Map<String, Type> validations, Logger logger) {
     bool isValid = true;
     for (final MapEntry<String, Object> entry in validations.entries) {
       if (!map.keys.contains(entry.key)) {
         isValid = false;
         logger.printTrace('The key `${entry.key}` was not found');
         break;
       }
-      if (map[entry.key] != null && (map[entry.key] as Object).runtimeType != entry.value) {
+      final Object? metadataValue = map[entry.key];
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/flutter_project_metadata.dart:62]: This was the bug: when `map[entry.key]` was null, this method was returning true (valid).
Instead, check the `map[entry.key]` `runtimeType` even when it is null so it is marked as invalid and prints `The value of key "x" in .metadata was expected to be y but was Null`

File: packages/flutter_tools/lib/src/flutter_project_metadata.dart [hunk 3]
```diff
@@ -89,28 +86,26 @@ class FlutterProjectMetadata {
     } on YamlException {
       // Handled in _validate below.
     }
-    if (yamlRoot == null || yamlRoot is! YamlMap) {
+    if (yamlRoot is! YamlMap) {
```
Comment (jmagman) [inline packages/flutter_tools/lib/src/flutter_project_metadata.dart:89]: `yamlRoot is! YamlMap` will be false for null, no need to check both.

**Review by christopherfujino — APPROVED** [review]
LGTM

**Pull request merged**